horovod/spark/keras/datamodule.py

```diff
@@ -0,0 +1,168 @@
+        # we choose is make_batch_reader unless there are data transformations.
+        self.transform_spec = TransformSpec(self.transform_fn) if self.transform_fn else None
+        self.reader_factory_kwargs = dict()
+        if self.transform_spec:
+            self.reader_factory = make_reader
+            self.reader_factory_kwargs['pyarrow_serialize'] = True
+            self.is_batch_reader = False
+        else:
+            self.reader_factory = make_batch_reader
+            self.is_batch_reader = True
+
+    def __enter__(self):
+        super().__enter__()
+        self.train_reader = self.reader_factory(self.train_dir,
+                                num_epochs=self.num_train_epochs,
+                                cur_shard=hvd.rank(),
+                                reader_pool_type=self.reader_pool_type,
+                                workers_count=self.train_reader_worker_count,
+                                shard_count=hvd.size(),
+                                hdfs_driver=constants.PETASTORM_HDFS_DRIVER,
+                                schema_fields=self.schema_fields,
+                                transform_spec=self.transform_spec,
+                                storage_options=self.storage_options,
+                                # Don't shuffle row groups if shuffle_buffer_size is 0 (non-shuffle case).
+                                shuffle_row_groups=True if self.shuffle_size > 0 else False,
+                                **self.reader_factory_kwargs)
+        self.val_reader = self.reader_factory(self.val_dir,
+                                num_epochs=1,
+                                cur_shard=hvd.rank(),
+                                reader_pool_type=self.reader_pool_type,
+                                workers_count=self.val_reader_worker_count,
+                                shard_count=hvd.size(),
+                                hdfs_driver=constants.PETASTORM_HDFS_DRIVER,
+                                schema_fields=self.schema_fields,
+                                transform_spec=self.transform_spec,
+                                storage_options=self.storage_options,
+                                shuffle_row_groups=False,
+                                **self.reader_factory_kwargs) if self.has_val else self.empty_batch_reader()
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self.has_val and self.val_reader:
+            self.val_reader.stop()
+        if self.train_reader:
+            self.train_reader.stop()
+        super().__exit__(type, value, traceback)
+
+    @contextlib.contextmanager
+    def empty_batch_reader(self):
+        yield None
+
+    def train_data(self):
+        return self.make_dataset(self.train_reader, self.train_batch_size, self.shuffle_size,
+                                 self.is_batch_reader, shuffle=True if self.shuffle_size > 0 else False,
```

Comment:
As mentioned outside, I would like to hold this PR till https://github.com/horovod/horovod/pull/3665 is merged due to Petastorm API changes.